music21/musicxml/m21ToXml.py

```diff
@@ -4421,9 +4487,33 @@ def noteToNotations(self, n, noteIndexInChord=0, chordParent=None):
             # get expressions from first note of chord
             chordOrNote = chordParent
 
-        # only apply expressions to notes or the first note of a chord...
-        if isSingleNoteOrFirstInChord:
-            for expObj in chordOrNote.expressions:
+        # apply all expressions apart from arpeggios only to the first note of a chord.
+        topNoteIndex: int
+        bottomNoteIndex: int
+        foundTopAndBottom = False
+        for expObj in chordOrNote.expressions:
+            mxExpression = None
+            if isinstance(expObj, expressions.ArpeggioMark):
+                if expObj.type == 'non-arpeggio':
+                    # <non-arpeggiate> goes on top and bottom note in chord
+                    if not foundTopAndBottom:
+                        topNoteIndex, bottomNoteIndex = self.topAndBottomIndices(chordOrNote)
+                        foundTopAndBottom = True
+                    if bottomNoteIndex == noteIndexInChord:
+                        mxExpression = self.expressionToXml(expObj)
+                        mxExpression.set('type', 'bottom')
+                    elif topNoteIndex == noteIndexInChord:
+                        mxExpression = self.expressionToXml(expObj)
+                        mxExpression.set('type', 'top')
+                else:
+                    # <arpeggiate> goes on every note in the chord
+                    mxExpression = self.expressionToXml(expObj)
+                if mxExpression is None:
+                    # the ArpeggioMark is not applicable on this note.
+                    continue
```

Comment:
Let's move all this (including the cached setup code) to its own arpeggioMarkToMxExpression method -- that way it can be tested separately and it doesn't interfere with reading this very important method.  It's a lot of extra lines to read through for the rather uncommon case of a ArpeggioMark let alone a Non-Arpeggiate mark.  Then it'll be easier to add future musicxml features as well.  (it's quite likely that this module is going to be split into smaller ones, like I started to do with `<sound>` and the refactor will be that easier too).

Reply:
done.